python/ray/train/lightning/lightning_trainer.py

```diff
@@ -0,0 +1,298 @@
+import pytorch_lightning as pl
+from pytorch_lightning.callbacks import ModelCheckpoint
+import logging
+
+from typing import TYPE_CHECKING, Callable, Dict, Optional, Union, Type, Any
+from inspect import isclass
+from pytorch_lightning.plugins.environments import ClusterEnvironment
+
+import ray
+from ray.air import session
+from ray.air.checkpoint import Checkpoint
+from ray.air.config import CheckpointConfig, DatasetConfig, RunConfig, ScalingConfig
+from ray.data.preprocessor import Preprocessor
+from ray.train.data_parallel_trainer import DataParallelTrainer
+from ray.train.torch.config import TorchConfig
+from ray.train.trainer import GenDataset
+from ray.train.constants import (
+    EVALUATION_DATASET_KEY,
+    TRAIN_DATASET_KEY,
+)
+from torch.utils.data import DataLoader
+from ray.train.lightning.lightning_utils import (
+    RayDDPStrategy, RayEnvironment, RayModelCheckpoint, RayDataModule
+)
+from ray.util import PublicAPI
+import ray.train as train
+
+logger = logging.getLogger(__name__)
+
+LIGHTNING_MODULE_KEY = "_lightning_module"
+LIGHTNING_MODULE_CONFIG_KEY = "_lightning_module_config"
+LIGHTNING_TRAINER_CONFIG_KEY = "_lightning_trainer_config"
+LIGHTNING_TRAINER_FIT_PARAMS_KEY = "_lightning_trainer_fit_params"
+MODEL_CHECKPOINT_CONFIG_KEY = "_model_checkpoint_config"
+DDP_STRATEGY_CONFIG_KEY = "_ddp_strategy_config"
+RAY_DATASET_ITER_CONFIG_KEY = "_ray_dataset_iter_config"
+
+
+@PublicAPI(stability="alpha")
+class LightningTrainer(DataParallelTrainer):
+    """A Trainer for data parallel PyTorch Lightning training.
+
+    This Trainer runs the ``pytorch_lightning.Trainer.fit()`` method on multiple
+    Ray Actors. The training is carried out in a distributed fashion through PyTorch
+    DDP. These actors already have the necessary Torch process group configured for 
+    distributed data parallel training.
+
+    The training function ran on every Actor will first initialize an instance
+    of the user-provided ``lightning_module`` class, which is a subclass of
+    ``pytorch_lightning.LightningModule`` using the arguments provided in
+    ``lightning_module_init_config``.
+
+    For data ingestion, the LightningTrainer will then either convert the Ray Dataset 
+    shards to a ``pytorch_lightning.LightningDataModule``, or directly use the datamodule 
+    if provided by users. 
+
+    The trainer will also create a ModelCheckpoint callback based on the configuration 
+    provided in ``model_checkpoint_config``. Notice that all the other ModelCheckpoint 
+    callbacks specified in ``lightning_trainer_config`` will be ignored.
+
+    Then, the training function will initialize an instance of ``pytorch_lightning.Trainer``
+    using the arguments provided in ``trainer_init_config`` and then run
+    ``pytorch_lightning.Trainer.fit``.
+
+    Args:
+        lightning_module: A subclass of ``pytorch_lightning.LightningModule``
```

Comment:
clarify that this should be the class definition and not the actual instance?

Reply:
👌